File: bootstrap/src/main/resources/executor/pi/src/pi-protocol-failure-receipt.mjs
import { createHash } from "node:crypto";
import { mkdir, rename, writeFile } from "node:fs/promises";
import { dirname } from "node:path";

import { canonicalizeStateV2 } from "./agent-state-v2-codec.mjs";

export const PI_PROTOCOL_FAILURE_RECEIPT_V1 = "PiProtocolFailureReceipt/v1";

export const PI_PROTOCOL_FAILURE_KINDS = Object.freeze({
  RESULT_MISSING_AFTER_RECOVERY: "RESULT_MISSING_AFTER_RECOVERY",
  AGENT_SETTLED_MISSING: "AGENT_SETTLED_MISSING",
  ROLE_SCHEMA_REJECTED_AFTER_RECOVERY: "ROLE_SCHEMA_REJECTED_AFTER_RECOVERY",
});

const RESULT_SOURCES = new Set(["NONE", "AGENT_RD_SUBMIT_RESULT", "BRIDGE_SYNTHETIC"]);
const REJECTION_KINDS = new Set(["NONE", "ROLE_SCHEMA"]);
const SHA256 = /^sha256:[0-9a-f]{64}$/;
const NON_EMPTY = /^[^\u0000-\u001f\u007f]+$/u;

export function canonicalizePiProtocolFailureReceipt(receipt) {
  validatePiProtocolFailureReceipt(receipt);
  return canonicalizeStateV2(receipt);
}

export function hashPiProtocolFailureReceipt(receipt) {
  return `sha256:${createHash("sha256")
    .update(canonicalizePiProtocolFailureReceipt(receipt), "utf8")
    .digest("hex")}`;
}

export async function writePiProtocolFailureReceipt(path, receipt) {
  const canonical = canonicalizePiProtocolFailureReceipt(receipt);
  await mkdir(dirname(path), { recursive: true });
  const temporary = `${path}.tmp-${process.pid}`;
  await writeFile(temporary, canonical, { encoding: "utf8", mode: 0o600 });
  await rename(temporary, path);
  return {
    canonical,
    hash: `sha256:${createHash("sha256").update(canonical, "utf8").digest("hex")}`,
  };
}

export function decodeAndVerifyPiProtocolFailureReceipt(json, expectedHash) {
  if (typeof json !== "string" || json.trim() === "") {
    throw new Error("receipt JSON must not be blank");
  }
  let receipt;
  try {
    receipt = JSON.parse(json);
  } catch (error) {
    throw new Error(`invalid receipt JSON: ${String(error.message ?? error)}`);
  }
  validatePiProtocolFailureReceipt(receipt);
  const canonical = canonicalizeStateV2(receipt);
  const actualHash = `sha256:${createHash("sha256").update(canonical, "utf8").digest("hex")}`;
  if (actualHash !== String(expectedHash ?? "").toLowerCase()) {
    throw new Error("receipt hash mismatch");
  }
  if (canonical !== json) {
    throw new Error("receipt JSON must use canonical bytes");
  }
  return receipt;
}

export function validatePiProtocolFailureReceipt(receipt) {
  if (!isPlainObject(receipt)) throw new Error("receipt must be an object");
  const expectedFields = [
    "acceptedResultDigest", "agentSettled", "attemptNo", "containerTerminated",
    "diagnosticArtifactIds", "eventStreamTrusted", "generatedAt", "kind",
    "lastRejectionDigest", "lastRejectionKind", "missingFacts", "protocol",
    "recoveryApplicable", "recoveryExhausted", "recoveryIssued", "resultSubmissionSource",
    "resultSubmitted", "role", "roleSchemaAccepted", "stageRunId", "taskId",
  ];
  const actualFields = Object.keys(receipt).sort();
  if (JSON.stringify(actualFields) !== JSON.stringify(expectedFields)) {
    throw new Error("receipt fields must exactly match v1 contract");
  }
  if (receipt.protocol !== PI_PROTOCOL_FAILURE_RECEIPT_V1) {
    throw new Error(`receipt protocol must be ${PI_PROTOCOL_FAILURE_RECEIPT_V1}`);
  }
  requireText(receipt.taskId, "taskId");
  requireText(receipt.stageRunId, "stageRunId");
  requireText(receipt.role, "role");
  if (!Number.isSafeInteger(receipt.attemptNo) || receipt.attemptNo < 1) {
    throw new Error("attemptNo must be a positive safe integer");
  }
  if (!PI_PROTOCOL_FAILURE_KINDS[receipt.kind]) throw new Error("unsupported receipt kind");
  if (!RESULT_SOURCES.has(receipt.resultSubmissionSource)) throw new Error("unsupported result submission source");
  if (!REJECTION_KINDS.has(receipt.lastRejectionKind)) throw new Error("unsupported last rejection kind");
  for (const field of [
    "resultSubmitted", "roleSchemaAccepted", "agentSettled", "eventStreamTrusted",
    "containerTerminated", "recoveryApplicable", "recoveryIssued", "recoveryExhausted",
  ]) {
    if (typeof receipt[field] !== "boolean") throw new Error(`${field} must be boolean`);
  }
  if (!Array.isArray(receipt.missingFacts) || receipt.missingFacts.some((item) => typeof item !== "string")) {
    throw new Error("missingFacts must be a string array");
  }
  if (!Array.isArray(receipt.diagnosticArtifactIds)
      || receipt.diagnosticArtifactIds.length === 0
      || receipt.diagnosticArtifactIds.some((item) => typeof item !== "string" || !NON_EMPTY.test(item))) {
    throw new Error("diagnosticArtifactIds must be a non-empty string array");
  }
  requireText(receipt.generatedAt, "generatedAt");
  if (!Number.isFinite(Date.parse(receipt.generatedAt))) throw new Error("generatedAt must be an instant");
  requireOptionalDigest(receipt.acceptedResultDigest, "acceptedResultDigest");
  requireOptionalDigest(receipt.lastRejectionDigest, "lastRejectionDigest");
  validateExactKindFacts(receipt);
  return receipt;
}

function validateExactKindFacts(receipt) {
  const facts = JSON.stringify(receipt.missingFacts);
  const recoveryAll = receipt.recoveryApplicable && receipt.recoveryIssued && receipt.recoveryExhausted;
  const noAcceptedAgentResult = receipt.resultSubmissionSource !== "AGENT_RD_SUBMIT_RESULT"
    && receipt.roleSchemaAccepted === false
    && receipt.acceptedResultDigest === "";
  if (receipt.kind === PI_PROTOCOL_FAILURE_KINDS.RESULT_MISSING_AFTER_RECOVERY) {
    if (facts !== '["AGENT_RESULT_SUBMITTED"]' || !receipt.agentSettled || !recoveryAll
        || !noAcceptedAgentResult || receipt.lastRejectionKind !== "NONE"
        || receipt.lastRejectionDigest !== "") {
      throw new Error("receipt contradicts RESULT_MISSING_AFTER_RECOVERY facts");
    }
    return;
  }
  if (receipt.kind === PI_PROTOCOL_FAILURE_KINDS.AGENT_SETTLED_MISSING) {
    if (facts !== '["AGENT_SETTLED"]' || receipt.agentSettled || !receipt.resultSubmitted
        || receipt.resultSubmissionSource !== "AGENT_RD_SUBMIT_RESULT" || !receipt.roleSchemaAccepted
        || !SHA256.test(receipt.acceptedResultDigest) || receipt.recoveryApplicable
        || receipt.recoveryIssued || receipt.recoveryExhausted || receipt.lastRejectionKind !== "NONE"
        || receipt.lastRejectionDigest !== "") {
      throw new Error("receipt contradicts AGENT_SETTLED_MISSING facts");
    }
    return;
  }
  if (facts !== '["AGENT_RESULT_SUBMITTED","ROLE_SCHEMA_ACCEPTED_RESULT"]'
      || !receipt.agentSettled || !recoveryAll || !noAcceptedAgentResult
      || receipt.lastRejectionKind !== "ROLE_SCHEMA" || !SHA256.test(receipt.lastRejectionDigest)) {
    throw new Error("receipt contradicts ROLE_SCHEMA_REJECTED_AFTER_RECOVERY facts");
  }
}

function requireText(value, field) {
  if (typeof value !== "string" || value.trim() === "" || !NON_EMPTY.test(value)) {
    throw new Error(`${field} must be non-blank text`);
  }
}

function requireOptionalDigest(value, field) {
  if (typeof value !== "string" || (value !== "" && !SHA256.test(value))) {
    throw new Error(`${field} must be blank or sha256 digest`);
  }
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

File: exec/src/main/java/com/wish/rd/exec/repair/pi/model/PiProtocolFailureReceipt.java
package com.wish.rd.exec.repair.pi.model;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.wish.rd.rag.project.agent.model.AgentStateV2Codec;

import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

/** Host-validated, canonical facts for the bounded Pi QA protocol retry. */
public record PiProtocolFailureReceipt(
        String protocol,
        Kind kind,
        List<String> missingFacts,
        boolean resultSubmitted,
        ResultSubmissionSource resultSubmissionSource,
        boolean roleSchemaAccepted,
        String acceptedResultDigest,
        boolean agentSettled,
        boolean eventStreamTrusted,
        boolean containerTerminated,
        boolean recoveryApplicable,
        boolean recoveryIssued,
        boolean recoveryExhausted,
        RejectionKind lastRejectionKind,
        String lastRejectionDigest,
        List<String> diagnosticArtifactIds,
        String taskId,
        String stageRunId,
        String role,
        int attemptNo,
        Instant generatedAt,
        String canonicalJson,
        String canonicalHash
) {

    public static final String PROTOCOL = "PiProtocolFailureReceipt/v1";
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    private static final Pattern SHA256 = Pattern.compile("sha256:[0-9a-f]{64}");
    private static final Set<String> EXACT_FIELDS = Set.of(
            "acceptedResultDigest", "agentSettled", "attemptNo", "containerTerminated",
            "diagnosticArtifactIds", "eventStreamTrusted", "generatedAt", "kind",
            "lastRejectionDigest", "lastRejectionKind", "missingFacts", "protocol",
            "recoveryApplicable", "recoveryExhausted", "recoveryIssued", "resultSubmissionSource",
            "resultSubmitted", "role", "roleSchemaAccepted", "stageRunId", "taskId"
    );

    public PiProtocolFailureReceipt {
        missingFacts = List.copyOf(missingFacts);
        diagnosticArtifactIds = List.copyOf(diagnosticArtifactIds);
    }

    public static PiProtocolFailureReceipt decodeAndVerify(String json, String expectedHash) {
        if (json == null || json.isBlank()) {
            throw new IllegalArgumentException("receipt JSON must not be blank");
        }
        JsonNode value;
        try {
            value = OBJECT_MAPPER.readTree(json);
        } catch (com.fasterxml.jackson.core.JsonProcessingException exception) {
            throw new IllegalArgumentException("invalid receipt JSON", exception);
        }
        if (value == null || !value.isObject()) {
            throw new IllegalArgumentException("receipt JSON must be an object");
        }
        String canonical = AgentStateV2Codec.canonicalize(value);
        String actualHash = AgentStateV2Codec.hash(value);
        if (!actualHash.equals(expectedHash == null ? "" : expectedHash.toLowerCase())) {
            throw new IllegalArgumentException("receipt hash mismatch");
        }
        if (!canonical.equals(json)) {
            throw new IllegalArgumentException("receipt JSON must use canonical bytes");
        }
        if (!EXACT_FIELDS.equals(fieldNames(value))) {
            throw new IllegalArgumentException("receipt fields must exactly match v1 contract");
        }
        if (!PROTOCOL.equals(value.path("protocol").asText())) {
            throw new IllegalArgumentException("receipt protocol must be " + PROTOCOL);
        }
        Kind kind = parseEnum(Kind.class, value.path("kind").asText(), "unsupported receipt kind");
        ResultSubmissionSource source = parseEnum(
                ResultSubmissionSource.class,
                value.path("resultSubmissionSource").asText(),
                "unsupported result submission source"
        );
        RejectionKind rejectionKind = parseEnum(
                RejectionKind.class,
                value.path("lastRejectionKind").asText(),
                "unsupported last rejection kind"
        );
        List<String> missingFacts = requiredStringList(value.get("missingFacts"), "missingFacts", false);
        List<String> artifactIds = requiredStringList(
                value.get("diagnosticArtifactIds"), "diagnosticArtifactIds", true
        );
        int attemptNo = value.path("attemptNo").asInt(0);
        if (!value.path("attemptNo").isIntegralNumber() || attemptNo < 1) {
            throw new IllegalArgumentException("attemptNo must be a positive safe integer");
        }
        Instant generatedAt;
        try {
            generatedAt = Instant.parse(requiredText(value, "generatedAt"));
        } catch (DateTimeParseException exception) {
            throw new IllegalArgumentException("generatedAt must be an instant", exception);
        }
        String acceptedDigest = optionalDigest(value, "acceptedResultDigest");
        String rejectionDigest = optionalDigest(value, "lastRejectionDigest");
        PiProtocolFailureReceipt receipt = new PiProtocolFailureReceipt(
                PROTOCOL,
                kind,
                missingFacts,
                requiredBoolean(value, "resultSubmitted"),
                source,
                requiredBoolean(value, "roleSchemaAccepted"),
                acceptedDigest,
                requiredBoolean(value, "agentSettled"),
                requiredBoolean(value, "eventStreamTrusted"),
                requiredBoolean(value, "containerTerminated"),
                requiredBoolean(value, "recoveryApplicable"),
                requiredBoolean(value, "recoveryIssued"),
                requiredBoolean(value, "recoveryExhausted"),
                rejectionKind,
                rejectionDigest,
                artifactIds,
                requiredText(value, "taskId"),
                requiredText(value, "stageRunId"),
                requiredText(value, "role"),
                attemptNo,
                generatedAt,
                json,
                expectedHash.toLowerCase()
        );
        validateExactKindFacts(receipt);
        return receipt;
    }

    public void requireIdentity(String expectedTaskId, String expectedStageRunId, String expectedRole, int expectedAttemptNo) {
        if (!taskId.equals(expectedTaskId)
                || !stageRunId.equals(expectedStageRunId)
                || !role.equals(expectedRole)
                || attemptNo != expectedAttemptNo) {
            throw new IllegalArgumentException("protocol failure receipt identity mismatch");
        }
    }

    public enum Kind {
        RESULT_MISSING_AFTER_RECOVERY,
        AGENT_SETTLED_MISSING,
        ROLE_SCHEMA_REJECTED_AFTER_RECOVERY
    }

    public enum ResultSubmissionSource {
        NONE,
        AGENT_RD_SUBMIT_RESULT,
        BRIDGE_SYNTHETIC
    }

    public enum RejectionKind {
        NONE,
        ROLE_SCHEMA
    }

    private static void validateExactKindFacts(PiProtocolFailureReceipt receipt) {
        boolean recoveryAll = receipt.recoveryApplicable && receipt.recoveryIssued && receipt.recoveryExhausted;
        boolean noAcceptedAgentResult = receipt.resultSubmissionSource != ResultSubmissionSource.AGENT_RD_SUBMIT_RESULT
                && !receipt.roleSchemaAccepted
                && receipt.acceptedResultDigest.isEmpty();
        switch (receipt.kind) {
            case RESULT_MISSING_AFTER_RECOVERY -> {
                if (!receipt.missingFacts.equals(List.of("AGENT_RESULT_SUBMITTED"))
                        || !receipt.agentSettled || !recoveryAll || !noAcceptedAgentResult
                        || receipt.lastRejectionKind != RejectionKind.NONE
                        || !receipt.lastRejectionDigest.isEmpty()) {
                    throw new IllegalArgumentException("receipt contradicts RESULT_MISSING_AFTER_RECOVERY facts");
                }
            }
            case AGENT_SETTLED_MISSING -> {
                if (!receipt.missingFacts.equals(List.of("AGENT_SETTLED"))
                        || receipt.agentSettled || !receipt.resultSubmitted
                        || receipt.resultSubmissionSource != ResultSubmissionSource.AGENT_RD_SUBMIT_RESULT
                        || !receipt.roleSchemaAccepted || !SHA256.matcher(receipt.acceptedResultDigest).matches()
                        || receipt.recoveryApplicable || receipt.recoveryIssued || receipt.recoveryExhausted
                        || receipt.lastRejectionKind != RejectionKind.NONE
                        || !receipt.lastRejectionDigest.isEmpty()) {
                    throw new IllegalArgumentException("receipt contradicts AGENT_SETTLED_MISSING facts");
                }
            }
            case ROLE_SCHEMA_REJECTED_AFTER_RECOVERY -> {
                if (!receipt.missingFacts.equals(List.of(
                        "AGENT_RESULT_SUBMITTED", "ROLE_SCHEMA_ACCEPTED_RESULT"
                )) || !receipt.agentSettled || !recoveryAll || !noAcceptedAgentResult
                        || receipt.lastRejectionKind != RejectionKind.ROLE_SCHEMA
                        || !SHA256.matcher(receipt.lastRejectionDigest).matches()) {
                    throw new IllegalArgumentException(
                            "receipt contradicts ROLE_SCHEMA_REJECTED_AFTER_RECOVERY facts"
                    );
                }
            }
        }
    }

    private static Set<String> fieldNames(JsonNode value) {
        java.util.HashSet<String> names = new java.util.HashSet<>();
        value.fieldNames().forEachRemaining(names::add);
        return Set.copyOf(names);
    }

    private static String requiredText(JsonNode value, String field) {
        JsonNode node = value.get(field);
        if (node == null || !node.isTextual() || node.textValue().isBlank()) {
            throw new IllegalArgumentException(field + " must be non-blank text");
        }
        return node.textValue();
    }

    private static boolean requiredBoolean(JsonNode value, String field) {
        JsonNode node = value.get(field);
        if (node == null || !node.isBoolean()) {
            throw new IllegalArgumentException(field + " must be boolean");
        }
        return node.booleanValue();
    }

    private static String optionalDigest(JsonNode value, String field) {
        JsonNode node = value.get(field);
        if (node == null || !node.isTextual()) {
            throw new IllegalArgumentException(field + " must be blank or sha256 digest");
        }
        String digest = node.textValue();
        if (!digest.isEmpty() && !SHA256.matcher(digest).matches()) {
            throw new IllegalArgumentException(field + " must be blank or sha256 digest");
        }
        return digest;
    }

    private static List<String> requiredStringList(JsonNode value, String field, boolean nonEmpty) {
        if (value == null || !value.isArray() || (nonEmpty && value.isEmpty())) {
            throw new IllegalArgumentException(field + " must be " + (nonEmpty ? "a non-empty " : "a ") + "string array");
        }
        java.util.ArrayList<String> values = new java.util.ArrayList<>();
        for (JsonNode item : value) {
            if (!item.isTextual() || item.textValue().isBlank()) {
                throw new IllegalArgumentException(field + " must be a string array");
            }
            values.add(item.textValue());
        }
        return List.copyOf(values);
    }

    private static <T extends Enum<T>> T parseEnum(Class<T> type, String value, String message) {
        try {
            return Enum.valueOf(type, value);
        } catch (IllegalArgumentException exception) {
            throw new IllegalArgumentException(message, exception);
        }
    }
}
